Backported by <stroboscope@de-bs.ru>

Author: Maxim Dounin <mdounin@mdounin.ru>

--- a/src/os/unix/ngx_readv_chain.c
+++ b/src/os/unix/ngx_readv_chain.c
@@ -10,7 +10,11 @@
 #include <ngx_event.h>
 
 
-#define NGX_IOVS  16
+#if (IOV_MAX > 64)
+#define NGX_IOVS 64
+#else
+#define NGX_IOVS IOV_MAX
+#endif
 
 
 #if (NGX_HAVE_KQUEUE)
@@ -71,7 +75,7 @@ ngx_readv_chain(ngx_connection_t *c, ngx
             iov->iov_len += chain->buf->end - chain->buf->last;
 
         } else {
-            if (vec.nelts >= IOV_MAX) {
+            if (vec.nelts >= NGX_IOVS) {
                 break;
             }
 
@@ -200,7 +204,7 @@ ngx_readv_chain(ngx_connection_t *c, ngx
             iov->iov_len += chain->buf->end - chain->buf->last;
 
         } else {
-            if (vec.nelts >= IOV_MAX) {
+            if (vec.nelts >= NGX_IOVS) {
                 break;
             }
 
